File: leetcode/0278_first_bad_version.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>

using namespace std;

bool isBadVersion(int version) {
    return version >= 3;
}

class Solution {
 public:
    int firstBadVersion(int n) {
        int mid, first = 1, last = n;

        while (first <= last) {
            mid = first + (last - first) / 2;
            if (isBadVersion(mid)) {
                last = mid - 1;
            } else {
                first = mid + 1;
            }
        }

        return first;
    }
};

int main() {
    auto solution = Solution();

    assert(3 == solution.firstBadVersion(5));
}

File: leetcode/0278_first_bad_version.py
# -*- coding: utf-8 -*-


def isBadVersion(n):
    return n >= 3


class Solution:
    def firstBadVersion(self, n):
        first, last = 1, n

        while first <= last:
            mid = (first + last) // 2
            if isBadVersion(mid):
                last = mid - 1
            else:
                first = mid + 1

        return first


if __name__ == '__main__':
    solution = Solution()

    assert 3 == solution.firstBadVersion(5)
